### PR TITLE
feat: add new encryption feature

### PR DESCRIPTION
**Pull request by shivamdurgbuns**
Closes # 

## 📑 Description

This PR will add the functionality where the API key will be encrypted before storing it in the config file.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## ℹ Additional Information

Added functionality where the key will be encrypted first and then added in the config file, and will be decrypted whenever it is used.

### CODE REVIEW

File: cmd/auth/auth.go [hunk 1]
```diff
@@ -26,6 +27,8 @@ var AuthCmd = &cobra.Command{
 	Long:  `Provide the necessary credentials to authenticate with your chosen backend.`,
 	Run: func(cmd *cobra.Command, args []string) {
 
+		var key = []byte{0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff, 0x00}
```
Comment (AlexsJones) [inline cmd/auth/auth.go:30]: I wonder if there could be another way to do this, like giving the user the option to set an encryption key or using something specific to them e.g. $WHOAMI etc

Comment (shivamdurgbuns) [reply cmd/auth/auth.go:30]: Would change this then

Comment (shivamdurgbuns) [reply cmd/auth/auth.go]: I have made the changes and added currentUser as the key

Comment (shivamdurgbuns) [reply cmd/auth/auth.go]: @AlexsJones could you please review this change and merge the PR

**Comment by thschue** [comment]
How does this work when the key is set programmatically (e.g. env or parameter)?

**Comment by shivamdurgbuns** [comment]
Looks like @thschue you have tagged this to a new issue, so will be working on this or someone else is taking care of it.

**Comment by shivamdurgbuns** [comment]
> How does this work when the key is set programmatically (e.g. env or parameter)?

Currently what I am doing is encrypting/decrypting the password before/after it is been written/read in the config file. So it should work even when the key is set programmatically as in the end the action will happen only after the value is fetched into/from config file. 
Looking at [#239](https://github.com/k8sgpt-ai/k8sgpt/issues/239) I think we might need to tinker around while reading from the password from env, as the config would only pass the env name and we would have to further read the value of the env before and after encryption. Password given by parameter is already covered.

Hope I am not missing something. Please correct me, if I am.

**Comment by AlexsJones** [comment]
Thanks for your contribution - I think overall there needs to be some thought into what this key is. 
If it's hard coded into the program it's going to be easy to see in the code and decrypt.

I feel there might be some option like...

```
k8sgpt auth lock --passphrase="foo"
```

```
k8sgpt auth unlock --passphrase="foo"
k8sgpt analyze
```

or 

```
k8sgpt auth lock --passphrase="foo"
k8sgpt analyze --passphase="foo"
```

What do you all think @thschue @shivamdurgbuns ?

**Comment by shivamdurgbuns** [comment]
@AlexsJones Sounds good to me, can add a new flag here, if @thschue agrees too.

**Comment by AlexsJones** [comment]
Okay I see reason in that - also less work to implement, go ahead.

**Comment by patrickpichler** [comment]
Do not get me wrong, but is it really worth it to implement locking/unlocking of the openai token
in the config? 

Imho leaving the job to a specialized tool (such as as a PW manager) or injecting it via an env variable 
is the "better" (highly subjective, so feel free to disagree :sweat_smile:) solution. It would also keep out
the complexity of keeping the unlocked state somewhere around.

File: cmd/auth/auth.go [hunk 2]
```diff
@@ -57,6 +69,7 @@ var AuthCmd = &cobra.Command{
 			os.Exit(1)
 		}
 
+		var encryptedPassword string
 		if password == "" {
```
[the remark this answers is not in the document]
Comment (shivamdurgbuns) [reply cmd/auth/auth.go:63]: I have add it likewise in the last commit. please review it.

File: cmd/auth/auth.go [hunk 3]
```diff
@@ -26,9 +28,19 @@ var AuthCmd = &cobra.Command{
 	Long:  `Provide the necessary credentials to authenticate with your chosen backend.`,
 	Run: func(cmd *cobra.Command, args []string) {
 
+		// Getting the user running the script to convert it as a key for encryption
+		currentUser, err := user.Current()
```
[the remark this answers is not in the document]
Comment (shivamdurgbuns) [reply cmd/auth/auth.go]: No It is not required, I have removed it.

File: cmd/analyze/analyze.go [hunk 4]
```diff
@@ -49,6 +49,8 @@ var AnalyzeCmd = &cobra.Command{
 		for _, provider := range configAI.Providers {
 			if backend == provider.Name {
 				aiProvider = provider
+				decryptedPassword := []byte(aiProvider.Password)
+				aiProvider.Password = string(decryptedPassword)
```
[the remark this answers is not in the document]
Comment (shivamdurgbuns) [reply cmd/analyze/analyze.go]: I added it for better readability, sure can update it.

File: pkg/lockandkey/encryption.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package lockandkey
```
[the remark this answers is not in the document]
Comment (shivamdurgbuns) [reply pkg/lockandkey/encryption.go:1]: Okay will update it like wise. I made it like that just to have a better separation.

**Comment by AlexsJones** [comment]
@shivamdurgbuns I wanted to thank you for the time you've taken to implement this and the patience you have shown.
I have doubts however that this is the right approach to take with regards to encryption. This specific implementation seems to leave out some important opportunities to use environmental variables or inject via password manager.

Therefore I will close this issue and invite you to participate in a new issue where we can discuss encryption at length before implementing something that will only last short term. 

Again, sorry if this feels like time wasted but I value your contribution and the time you have taken to improve our ecosystem - please do help us with the next steps

**Comment by shivamdurgbuns** [comment]
No problem! Whatever is good for the project :)

**Pull request closed without merging**